Theory-on-regression.md

```diff
@@ -0,0 +1,97 @@
+
+$$ \color{gray} \text{repeat until convergence: } \\
+     \color{black} w_i := w_i - \alpha\frac{\partial{}}{\partial{w}}J(w) \\
+    \color{gray}
+    \text{for } i = (0, \cdots, n) \\
+    \alpha \text{ is learning rate}
+$$
+
+The basic idea is to take repeated steps in the opposite direction of the gradient of the loss function, i.e. $\displaystyle -\frac{\partial{}}{\partial{w}}J(w)$, which will lead to a local minimum of loss function, as shown below.
+
+![Gradient descent 1D](image/Gradient-descent.png)
+<!-- Figure Gradient-descent.png from https://imaddabbura.github.io/img/gradient-descent-algorithms/gradients.PNG -->
+
+The process of gradient descent for two features would look like the dynamic diagram below:
+![Gradient descent 2D dynamic](image/Gradient_descent_dynamic.gif)
+<!-- Figure Gradient_descent_dynamic.gif from https://blog.paperspace.com/intro-to-optimization-in-deep-learning-gradient-descent/ -->
+
+
+## Neural network (non-linear regression)
+
+### Model representation
+
+![Neural network](image/Neural_network.png)
+
+The diagram demonstrates one hidden layer neural network.
+
+<!-- Text from https://scikit-learn.org/stable/modules/neural_networks_supervised.html#multi-layer-perceptron -->
+The input layer consists of a set of neurons representing the input features.
+
+ Each neuron in the hidden layer transforms the values from the previous layer with a weighted linear summation $w_1x_1 + w_2x_2 + \cdots + w_nx_n + bias$, followed by a non-linear activation function, e.g. the [logistic or sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function).
+
+ The output layer receives the values from the last hidden layer and transforms them into output values with linear function.
+
+The neural network available in `scikit-learn` is just look like the architecture in the diagram, i.e. multi-layer perceptron. To use more complex neural network models, deep learning framework should be used, e.g. [PyTorch, Keras, TensorFlow, etc](https://en.wikipedia.org/wiki/Comparison_of_deep_learning_software). Want to have a look at various neural network models? Try https://alexlenail.me/NN-SVG/index.html.
+
+#### Activation function
+[Activation function](https://en.wikipedia.org/wiki/Activation_function) is also called transfer function. Various activation functions are available in `scikit-learn`, e.g. identity, logistic, tanh and relu. Take a look at [this table](https://en.wikipedia.org/wiki/Activation_function#Table_of_activation_functions) to know more about these functions.
+
+![Activation functions](image/Activation_functions.png)
+<!-- Table is from https://en.wikipedia.org/wiki/Activation_function#Table_of_activation_functions -->
+
+
+### Loss function
+Loss function can be shared by all regression models, so loss function $J(w) = ||y - \hat{y}||^2$ is also used for neural network model.
+
+### Optimization algorithms
```

Comment:
Here I also don't think you have to go into too much detail about different optimizers. It could be useful to show an image of backpropegation.

Reply:
Yes, same as the activation function, we won't mention too much detail about optimisers, but the commonly used choice.
As for the backprop, I even don't want to mention it actually. No more image, otherwise we have to explain more...